### ISSUE TITLE
BUG: unable to handle page fault for address: ffff8cb1a7beb900

### ISSUE THREAD

**Issue by KyleSanderson** [trapexit/mergerfs#686]
General description
===================
MergerFS seems to trigger a ton of FUSE BUGs it seems... here's one.

Expected behavior
=================
Files served.

Actual behavior (system info below).
===============
```
[ 3517.458859] kernel tried to execute NX-protected page - exploit attempt? (uid: 0)
[ 3517.458955] BUG: unable to handle page fault for address: ffff8cb1a7beb900
[ 3517.459044] #PF: supervisor instruction fetch in kernel mode
[ 3517.459130] #PF: error_code(0x0011) - permissions violation
[ 3517.459216] PGD 468e01067 P4D 468e01067 PUD 6cd929063 PMD 8000000727a001e3
[ 3517.459308] Oops: 0011 [#1] PREEMPT SMP NOPTI
[ 3517.459394] CPU: 3 PID: 4965 Comm: mergerfs Not tainted 5.4.6-gentoo #1
[ 3517.459481] Hardware name: Supermicro Super Server/A2SDi-8C-HLN4F, BIOS 1.1c 06/25/2019
[ 3517.459575] RIP: 0010:0xffff8cb1a7beb900
[ 3517.459659] Code: ff ff d0 b8 be a7 b1 8c ff ff e0 b8 be a7 b1 8c ff ff e0 b8 be a7 b1 8c ff ff 30 25 f8 6c b2 8c ff ff 00 00 00 00 00 00 00 00 <00> 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 e0 2d 1a d7 b2 8c
[ 3517.459771] RSP: 0018:ffffa37381b87c88 EFLAGS: 00010286
[ 3517.459856] RAX: ffff8cb1a7beb900 RBX: ffff8cb2d5539088 RCX: 0000000000000001
[ 3517.459945] RDX: 0000000000000000 RSI: ffffa3738047bdc0 RDI: ffff8cb2d376ba00
[ 3517.460034] RBP: ffff8cb2d376ba00 R08: 0000000000000000 R09: ffffa37381b87c28
[ 3517.460123] R10: ffffa37381b87b8c R11: 0000000000000000 R12: ffff8cb2d5539098
[ 3517.460212] R13: ffff8cb1a7beb900 R14: ffffa37381b87d28 R15: ffff8cb2d7572cc0
[ 3517.460302] FS:  00007f9a51cf5700(0000) GS:ffff8cb2dfac0000(0000) knlGS:0000000000000000
[ 3517.460838] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 3517.461147] CR2: ffff8cb1a7beb900 CR3: 0000000852c56000 CR4: 00000000003406e0
[ 3517.461457] Call Trace:
[ 3517.461774]  ? fuse_request_end+0xae/0x1c0 [fuse]
[ 3517.462088]  ? fuse_dev_do_write+0x2a0/0x3e0 [fuse]
[ 3517.462403]  ? fuse_dev_write+0x6e/0xa0 [fuse]
[ 3517.462715]  ? do_iter_readv_writev+0x150/0x1c0
[ 3517.463023]  ? do_iter_write+0x80/0x190
[ 3517.463329]  ? vfs_writev+0xa3/0x100
[ 3517.463639]  ? do_vfs_ioctl+0xa4/0x620
[ 3517.463946]  ? __fget+0x73/0xb0
[ 3517.464251]  ? do_writev+0x65/0x100
[ 3517.464559]  ? do_syscall_64+0x54/0x190
[ 3517.464866]  ? entry_SYSCALL_64_after_hwframe+0x49/0xbe
[ 3517.465176] Modules linked in: nfsd auth_rpcgss nfs_acl tda18271 s5h1411 cfg80211 rfkill 8021q garp stp mrp llc xt_length xt_conntrack ip6table_filter ip6_tables cachefiles x86_pkg_temp_thermal nf_conntrack_ftp kvm_intel nf_conntrack nf_defrag_ipv4 kvm irqbypass snd_pcm coretemp saa7164 snd_timer crct10dif_pclmul tveeprom bcache snd crc32_pclmul crc64 dvb_core crc32c_intel soundcore ghash_clmulni_intel pcspkr videodev mc i2c_ismt acpi_cpufreq xts aesni_intel crypto_simd cryptd glue_helper crc32_generic cbc sha256_generic libsha256 ixgb ixgbe tulip cxgb3 cxgb mdio cxgb4 vxge bonding vxlan ip6_udp_tunnel udp_tunnel macvlan vmxnet3 tg3 sky2 r8169 pcnet32 mii igb dca i2c_algo_bit i2c_core e1000 bnx2 atl1c msdos fat cramfs squashfs fuse xfs nfs lockd grace sunrpc fscache jfs reiserfs btrfs zstd_decompress zstd_compress ext4 jbd2 ext2 mbcache linear raid10 raid1 raid0 dm_zero dm_verity reed_solomon dm_thin_pool dm_switch dm_snapshot dm_raid raid456 async_raid6_recov async_memcpy async_pq
[ 3517.465251]  raid6_pq dm_mirror dm_region_hash dm_log_writes dm_log_userspace dm_log dm_integrity async_xor async_tx xor dm_flakey dm_era dm_delay dm_crypt dm_cache_smq dm_cache dm_persistent_data libcrc32c dm_bufio dm_bio_prison dm_mod dax firewire_core crc_itu_t sl811_hcd xhci_pci xhci_hcd usb_storage mpt3sas raid_class aic94xx libsas lpfc nvmet_fc nvmet qla2xxx megaraid_sas megaraid_mbox megaraid_mm aacraid sx8 hpsa 3w_9xxx 3w_xxxx 3w_sas mptsas scsi_transport_sas mptfc scsi_transport_fc mptspi mptscsih mptbase imm parport sym53c8xx initio arcmsr aic7xxx aic79xx scsi_transport_spi sr_mod cdrom sg sd_mod pdc_adma sata_inic162x sata_mv ata_piix ahci libahci sata_qstor sata_vsc sata_uli sata_sis sata_sx4 sata_nv sata_via sata_svw sata_sil24 sata_sil sata_promise pata_via pata_jmicron pata_marvell pata_sis pata_netcell pata_pdc202xx_old pata_atiixp pata_amd pata_ali pata_it8213 pata_pcmcia pata_serverworks pata_oldpiix pata_artop pata_it821x pata_hpt3x2n pata_hpt3x3 pata_hpt37x pata_hpt366
[ 3517.468482]  pata_cmd64x pata_sil680 pata_pdc2027x nvme_fc nvme_rdma rdma_cm iw_cm ib_cm ib_core ipv6 crc_ccitt nf_defrag_ipv6 configfs nvme_fabrics virtio_net net_failover failover virtio_crypto crypto_engine virtio_mmio virtio_pci virtio_balloon virtio_rng virtio_console virtio_blk virtio_scsi virtio_ring virtio
[ 3517.472664] CR2: ffff8cb1a7beb900
[ 3517.472976] ---[ end trace 526b6506602bdb74 ]---
[ 3517.516624] RIP: 0010:0xffff8cb1a7beb900
[ 3517.516964] Code: ff ff d0 b8 be a7 b1 8c ff ff e0 b8 be a7 b1 8c ff ff e0 b8 be a7 b1 8c ff ff 30 25 f8 6c b2 8c ff ff 00 00 00 00 00 00 00 00 <70> 2b 40 e7 ad 8c ff ff 60 89 84 b4 ff ff ff ff e0 2d 1a d7 b2 8c
[ 3517.517745] RSP: 0018:ffffa37381b87c88 EFLAGS: 00010286
[ 3517.518055] RAX: ffff8cb1a7beb900 RBX: ffff8cb2d5539088 RCX: 0000000000000001
[ 3517.518368] RDX: 0000000000000000 RSI: ffffa3738047bdc0 RDI: ffff8cb2d376ba00
[ 3517.518681] RBP: ffff8cb2d376ba00 R08: 0000000000000000 R09: ffffa37381b87c28
[ 3517.518994] R10: ffffa37381b87b8c R11: 0000000000000000 R12: ffff8cb2d5539098
[ 3517.519308] R13: ffff8cb1a7beb900 R14: ffffa37381b87d28 R15: ffff8cb2d7572cc0
[ 3517.519622] FS:  00007f9a51cf5700(0000) GS:ffff8cb2dfac0000(0000) knlGS:0000000000000000
[ 3517.520159] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 3517.520469] CR2: ffff8cb1a7beb900 CR3: 0000000852c56000 CR4: 00000000003406e0
```
```
mergerfs version: 2.28.3-4-g55245e2
FUSE library version: 2.9.7-mergerfs_2.29.0
fusermount version: 2.9.8
using FUSE kernel interface version 7.29
```

**Comment by trapexit** [comment]
We'll see what the kernel maintainers say. This is outside my domain.

**Comment by aqxa1** [comment]
I'm having the same issue (5.4.8 kernel). Currently, I'm testing 5.3.18 to see if the issue exists in the older kernel.

**Comment by trapexit** [comment]
@KyleSanderson brought it up on the linux-fsdevel mailinglist but hasn't replied to the response nor has Miklos replied.

**Comment by KyleSanderson** [comment]
@trapexit do you mean to Greg? I can't run older kernels because of the xfs features I'm using.

**Comment by trapexit** [comment]
Yes.  You should probably reply regardless.

What feature are you using?

**Comment by KyleSanderson** [comment]
Fixed - https://git.kernel.org/pub/scm/linux/kernel/git/mszeredi/fuse.git/commit/?h=for-next&id=3e8cb8b2eaeb22f540f1cbc00cbb594047b7ba89

**Comment by KyleSanderson** [comment]
OK so - the regression came in with 5.4.0-rc(0) / up to (maybe including 5.4.23) are impacted. 5.5 is going to be forever broken. This is still broken in https://github.com/torvalds/linux/blob/63623fd44972d1ed2bfb6e0fb631dfcf547fd1e7/fs/fuse/dev.c#L279 .

It's in fuse-next so that's progress in the right direction. Can you add a notice? This is going to become more apparent as distros continue to greenlight 5.4.

**Comment by trapexit** [comment]
I imagine the distros will wait for this to make it into the stable branch or have to otherwise manage this given mergerfs is hardly the only FUSE filesystem. For instance gvfs used on GTK/GNOME platforms.

**Comment by KyleSanderson** [comment]
Yeah, low risk currently outside of folk running unstable distros.  Right now it's just going to hit testers mostly. Ubuntu is on this schedule
```Testing week: January 9, 2020
UI Freeze: March 19, 2020
Ubuntu 20.04 Beta: April 2, 2020
Kernel Freeze: April 9, 2020
Release Candidate: April 16, 2020
```

Fedora 31 already has 5.4.20 marked as stable - which will hit this.
Gentoo still has 5.4.22 (5.4 entire series) soft-masked.

**Comment by rfjakob** [comment]
Hi FUSE friends, I have gocryptfs users [hitting this as well](https://github.com/rfjakob/gocryptfs/issues/458).

@KyleSanderson why do you think kernel 5.5 is not going to get the patch? Fedora 31 is on 5.5.6 already, so the 5.5 series not getting the patch would be pretty sad.

**Comment by trapexit** [comment]
I don't think 5.5 will be a longterm version. 5.4 is. That said... any major distro will likely pick up the patch. Too serious IMO not to do so.

**Comment by KyleSanderson** [comment]
@rfjakob this still is not upstream in 5.4.24... I know the majority of kernel maintainers see fuse as a toy but it's pretty bad that the fix is still not in the stable tree (I did send a note to greg/similar on fuse-devel confirming the patch fixes it)...

The patch was tagged for stable: no idea when it's going to get picked up. 5.5 is not stable, and will likely not get the patch regardless.

**Comment by rfjakob** [comment]
Linux v5.5.13 now has the patch: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/log/?h=v5.5.13&qt=author&q=Szeredi

So does Linux v5.4.28: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/log/?h=v5.4.28&qt=author&q=Szeredi

Yay!

**Comment by rfjakob** [comment]
Apparently the fix got in in earlier: Linux v5.5.10 and v5.4.26 already have it.

Found out using (commit ids from the links above):

```
$ git describe --contains 2031a50a17f2f3b99d5fc9b6b8fe85688b139a5b
v5.5.10~51
```

```
$ git describe --contains 63050b3dc02f2bbf1a1770a11a6fe1c157645df5
v5.4.26~44
```

**Comment by KyleSanderson** [comment]
Yup, ACK. This appears to be fixed now upstream (finally!).
